### PR TITLE
graph: add parent img refcount for faster rmi

### LINKED ISSUE #16771: Terrible `docker rmi` performance
Currently, a daemon with >50 images takes a ridiculous amount of time to honor a single `rmi` request.

After investigation, the root cause of the issue is [this line](https://github.com/docker/docker/blob/ec1d8d7f2a8f563deb61df2b35aa8a9d6a07c3e2/daemon/image_delete.go#L282). The `HasChildren` function is incredibly complex: it calls into [`graph.ByParent()`](https://github.com/docker/docker/blob/ec1d8d7f2a8f563deb61df2b35aa8a9d6a07c3e2/graph/graph.go#L391) that walks the entire graph just for the sake of determining if an image has children.

Having a "fast-exit" (e.g., stop the loop when a first children is found) won't solve the most common case. We should look into maybe removing that `ByParent()` function, and maintain a reference count for each graph entry such as delete would fail immediately unless the counter is down to 0.

I would love to fix this for 1.9.0 but still haven't found the time to do it myself :-( Ping @LK4D4 @runcom if interested!

PS: the issue is easily reproduced by doing a `docker pull --all-tags` on `ubuntu` or `debian`.

### PR DESCRIPTION
**Pull request by runcom**
ping @LK4D4 @icecrime 

Fixes #16771

Signed-off-by: Antonio Murdaca amurdaca@redhat.com

### CODE REVIEW

File: pkg/truncindex/truncindex.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Package truncindex package provides a general 'index tree', used by Docker
+// Package truncindex provides a general 'index tree', used by Docker
```
Comment (LK4D4) [inline pkg/truncindex/truncindex.go]: note this in commit mesage

File: graph/graph.go [hunk 2]
```diff
@@ -82,6 +82,9 @@ type Graph struct {
 	imageMutex       imageMutex // protect images in driver.
 	retained         *retainedLayers
 	tarSplitDisabled bool
+
+	parentRefs      map[string]uint16
```
Comment (LK4D4) [inline graph/graph.go]: So, if we want to be superoptimized, then let's just use `uint32` and `atomic.AddUint32` without mutex

Comment (runcom) [inline graph/graph.go]: cool

File: graph/graph.go [hunk 3]
```diff
@@ -359,6 +374,13 @@ func (graph *Graph) Delete(name string) error {
 	}
 	// Remove rootfs data from the driver
 	graph.driver.Remove(id)
+
+	graph.parentRefsMutex.Lock()
+	if img.Parent != "" {
+		graph.parentRefs[img.Parent]--
```
Comment (LK4D4) [inline graph/graph.go]: You need `delete` if it's 0 here.

**Comment by icecrime** [comment]
Huge performance improvement ;-) LGTM!

**Comment by runcom** [comment]
*fuch* *fuch* *master*

**Comment by cpuguy83** [comment]
I think this could be used to speed up build cache lookup as well.
The build cache can ask for the previous layer's children and first check if any of the children are matches to the current dockerfile line rather walking the whole graph first.

**Comment by LK4D4** [comment]
@cpuguy83 makes sense

**Comment by LK4D4** [comment]
LGTM

**Comment by coolljt0725** [comment]
I think we should revert this. with this `docker images` show all the images without `-a`
image

and we can remove the image even though it has children
for example the image `ae71e44e793e` is the children of `012f682ad4c8`
<pre><code>[lei@fedora docker]$ docker inspect ae71e44e793e
[
{
    "Id": "ae71e44e793ec9d6aa5e1983eeb7e9e10782b9b55efd06823fb7f5356a91d4e3",
    "Tags": [],
    "Parent": "012f682ad4c80d5bdbd709e75d8920136d49025cefb7a1a38fab2a9acbd76caa",
</code></pre>
but I can still remove `012f682ad4c8`
<pre><code>[lei@fedora docker]$ docker rmi 012f682ad4c8
Deleted: 012f682ad4c80d5bdbd709e75d8920136d49025cefb7a1a38fab2a9acbd76caa
Deleted: 386ddb3303b83b32674c3f43173f689ee23ef41b54e511c43175b2530576be5e
Deleted: 2ad11c25e421aaaebc9a0644125e69ecc04ef95d4abdc2baec4ec29df9aae7b7
Deleted: b519c4d3055400bace81553d0db8182c3e1ac7182fa73c5f215f5a79225ab1df
Deleted: c0767f27abaf04c4837cac8fbbd68f800bb2a0c35158ea7f7ffd0afdfb7172ed
Deleted: b8cce848b30d56a17b8d4f554771d7c76bb4f49289cef167f11eb1f5260c0822
Deleted: 9697b7ab494a1a5c097b019c74f2901c5185c532e03cd73f48af61711df6e538
Deleted: 029c347defbab365fda45afc3290bf41e507342b17daadca9f41de9e91ac5707
Deleted: c6ab4dc7ed3049fc3d25977d52c792cf42d5d2b2a4a9ab417d9a65e9d8283108
Deleted: 5bd12ced85aee86c2299b3f5390b0c677294fa7168630b286903d0fee57fb9d0
Deleted: b33a4411cc4145bb24388c430d185bb0beb8b2814b69d6e1b88be5b09269ce42
Deleted: be30f5a8d9405ef0991ce91ce609e9bfa8c11e0be5e56ded6120e1c3b4d476c4
Deleted: 00b6cb4c5b9b15425a79750e46b125bb01f45af08cbb817cd67ce97529582946
Deleted: 085dcd89c60bc4a762a2b11c12b8481b994a2a2e5528d68cd911b1d6136e8136
Deleted: 7094222eee30293333fa5ef74885e0c50c2a18b0d49fba8cce540407ec097e03
Deleted: 9213e81cb0f2a81133c82bc93394b600f532866e3aaf78d9c5dee79bedc64704
Deleted: 607e965985c11e6a23270feec487908aeaa9af763d24a2986866a41537770c8c
Deleted: 1ff9f26f09fb1bc7b5955c269b1042429e86d7891c653f52f3e48f1e0365d7df
Deleted: 9a61b6b1315e6b457c31a03346ab94486a2f5397f4a82219bee01eead1c34c2e
Deleted: 902b87aaaec929e80541486828959f14fa061f529ad7f37ab300d4ef9f3a0dbf</code></pre>

**Pull request merged**